### PR TITLE
Change integer types to scale up on 64 bit architectures

### LINKED ISSUE #145: Fixup integer types
There are some cases where a u32 is used but a usize is better as it would scale up on a 64 bit architecture. These should be changed to make the code more versatile.

Found culprits are:
* `kmap_end` and `kmap_start` in panic.init should be declared as usize
* `mod_start` and `mod_end` in panic.init should be casted to a usize
* `arch.MEMORY_BLOCK_SIZE`, `pmm.BLOCK_SIZE` and `vmm.BLOCK_SIZE` should be usize
* `mem.MemProfile.mem_kb`
* `Bitmap.init` should accept the number of entries as a usize
* Bitmap types should use usize as the entry type

### PR DESCRIPTION
**Pull request by SamTebbs33**
Closes #145

### CODE REVIEW

File: src/kernel/vmm.zig [hunk 1]
```diff
@@ -161,7 +161,7 @@ pub fn VirtualMemoryManager(comptime Payload: type) type {
         ///
         pub fn init(start: usize, end: usize, allocator: *std.mem.Allocator, mapper: Mapper(Payload), payload: Payload) std.mem.Allocator.Error!Self {
             const size = end - start;
-            var bmp = try bitmap.Bitmap(u32).init(@floatToInt(u32, @ceil(@intToFloat(f32, size) / @intToFloat(f32, pmm.BLOCK_SIZE))), allocator);
+            var bmp = try bitmap.Bitmap(usize).init(std.mem.alignForward(size, pmm.BLOCK_SIZE) / pmm.BLOCK_SIZE, allocator);
```
Comment (DrDeano) [inline src/kernel/vmm.zig:164]: I like this :)

File: src/kernel/vmm.zig [hunk 2]
```diff
@@ -259,7 +259,7 @@ pub fn VirtualMemoryManager(comptime Payload: type) type {
         /// Error: std.mem.Allocator.Error
         ///     std.mem.AllocatorError.OutOfMemory: The required amount of memory couldn't be allocated
         ///
-        pub fn alloc(self: *Self, num: u32, attrs: Attributes) std.mem.Allocator.Error!?usize {
+        pub fn alloc(self: *Self, num: usize, attrs: Attributes) std.mem.Allocator.Error!?usize {
```
Comment (DrDeano) [inline src/kernel/vmm.zig:262]: Update the doc comment :P

File: src/kernel/pmm.zig [hunk 3]
```diff
@@ -88,7 +88,7 @@ pub fn free(addr: usize) (PmmBitmap.BitmapError || PmmError)!void {
 /// Return: u32.
 ///     The number of unallocated blocks of memory
 ///
-pub fn blocksFree() u32 {
+pub fn blocksFree() usize {
```
Comment (DrDeano) [inline src/kernel/pmm.zig:91]: Doc comment

File: src/kernel/bitmap.zig [hunk 4]
```diff
@@ -215,7 +215,7 @@ pub fn Bitmap(comptime BitmapType: type) type {
         /// Error: BitmapError.
         ///     OutOfBounds: The index given is out of bounds.
         ///
-        pub fn isSet(self: *const Self, idx: u32) BitmapError!bool {
+        pub fn isSet(self: *const Self, idx: usize) BitmapError!bool {
```
Comment (DrDeano) [inline src/kernel/bitmap.zig:218]: Doc comment

File: src/kernel/bitmap.zig [hunk 5]
```diff
@@ -188,13 +188,13 @@ pub fn Bitmap(comptime BitmapType: type) type {
         ///     The index within all bitmaps that was set or null if there wasn't one free.
         ///     0 .. ENTRIES_PER_BITMAP - 1 if in the first bitmap, ENTRIES_PER_BITMAP .. ENTRIES_PER_BITMAP * 2 - 1 if in the second etc.
         ///
-        pub fn setFirstFree(self: *Self) ?u32 {
+        pub fn setFirstFree(self: *Self) ?usize {
```
Comment (DrDeano) [inline src/kernel/bitmap.zig:191]: Doc comment

File: src/kernel/bitmap.zig [hunk 6]
```diff
@@ -49,8 +49,8 @@ pub fn Bitmap(comptime BitmapType: type) type {
         /// Error: std.mem.Allocator.Error
         ///     OutOfMemory: There isn't enough memory available to allocate the required number of BitmapType.
         ///
-        pub fn init(num_entries: u32, allocator: *std.mem.Allocator) !Self {
-            const num = @floatToInt(u32, @ceil(@intToFloat(f32, num_entries) / @intToFloat(f32, ENTRIES_PER_BITMAP)));
+        pub fn init(num_entries: usize, allocator: *std.mem.Allocator) !Self {
```
Comment (DrDeano) [inline src/kernel/bitmap.zig:52]: Doc comment

File: src/kernel/arch/x86/paging.zig [hunk 7]
```diff
@@ -99,10 +99,10 @@ const TENTRY_AVAILABLE: u32 = 0xE00;
 const TENTRY_PAGE_ADDR: u32 = 0xFFFFF000;
 
 /// The number of bytes in 4MB
-pub const PAGE_SIZE_4MB: u32 = 0x400000;
+pub const PAGE_SIZE_4MB: usize = 0x400000;
 
 /// The number of bytes in 4KB
-pub const PAGE_SIZE_4KB: u32 = PAGE_SIZE_4MB / 1024;
+pub const PAGE_SIZE_4KB: usize = PAGE_SIZE_4MB / 1024;
```
Comment (DrDeano) [inline src/kernel/arch/x86/paging.zig:105]: Do these need to be `usize` as this in the x86

File: src/kernel/vmm.zig (hunk 1 above)
Comment (SamTebbs33) [reply src/kernel/vmm.zig:164]: Me too :D I don't know why I didn't do it this way in the first place

File: src/kernel/arch/x86/paging.zig (hunk 7 above)
Comment (SamTebbs33) [reply src/kernel/arch/x86/paging.zig:105]: I guess it doesn't need to be for the reasons I listed in the issue, but I think that something describing a size should be `usize` anyway. It also means casting is unnecessary when using it as the block size for the vmm and pmm.

**Pull request merged**